File: FalsePositionMethod.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Function for which we want to find the root
double fx(double x) {
    return (x * x * x) - (2 * x) - 5; // Example: f(x) = x^3 - 2x - 5
}

// Implementing the False Position Method
double falsi(double x0, double x1, int maxitr, float e) {
    if (fx(x0) * fx(x1) > 0) {
        cout<<"Wrong guess values."<<endl;
        return 0;
    }

    double x2;
    for (int i = 0; i < maxitr; i++) {
        // Calculate the new approximation using the False Position formula
        x2 = x0 - ((x1 - x0) * fx(x0)) / (fx(x1) - fx(x0));

        // Print the current iteration and the value of x2
        cout << "Iteration " << i + 1 << " x2 = " << x2 << " fx(x2) = " << fx(x2) << endl;

        // Check if the result is within the desired tolerance
        // fabs() function is in <cmath> file whis gives |absolute| value 
        if (fabs(fx(x2)) <= e) {
            cout << "The root is: " << x2 << endl;
            return x2;
        }

        // Update x0 or x1 based on the sign of f(x2)
        if (fx(x2) * fx(x0) < 0)
            x1 = x2;
        else
            x0 = x2;
    }

    cout << "The method did not converge after " << maxitr << " iterations." << endl;
    return 0;
}

int main() {
    double x0, x1;
    int maxitr;
    float e = 0.0001; // Tolerable error

    cout << "Enter first guess value (x0): " << endl;
    cin >> x0;
    cout << "Enter second guess value (x1): " << endl;
    cin >> x1;
    cout << "Enter the number of iterations: " << endl;
    cin >> maxitr;

    // Call the false position method
    falsi(x0, x1, maxitr, e);

    return 0;
}

File: FalsePositionMethod.py
import math

# Function to evaluate f(x) = x^3 - 2x - 5
def fx(x):
    return (x * x * x) - (2 * x) - 5

# False position (Regula Falsi) method
def falsi(x0, x1, maxitr, e):
    # Check if the initial guesses do not bracket the root
    if fx(x0) * fx(x1) > 0:
        print("Wrong guess values.")
        return None

    x2 = 0
    for i in range(maxitr):
        # Compute the new approximation for the root
        x2 = x0 - ((x1 - x0) * fx(x0)) / (fx(x1) - fx(x0))

        # Print the current iteration, approximation, and function value
        print(f"Iteration {i + 1}: x2 = {x2}, f(x2) = {fx(x2)}")

        # If the function value at the new approximation is within tolerance, stop
        if abs(fx(x2)) <= e:
            print(f"The root is: {x2}")
            return x2

        # Decide which subinterval to use in the next step
        if fx(x2) * fx(x0) < 0:
            x1 = x2
        else:
            x0 = x2

    print(f"The method did not converge after {maxitr} iterations.")
    return None

# Input values from the user
x0 = float(input("Enter first guess value (x0): "))
x1 = float(input("Enter second guess value (x1): "))
maxitr = int(input("Enter the number of iterations: "))
e = float(input("Enter the tolerable error (e.g., 0.00001): "))

# Call the falsi method
falsi(x0, x1, maxitr, e)
